Let's try this again

Paladin opened the pull request:
Thought I'd submitted this pull request earlier today, but I don't see it anywhere, so maybe I fatfingered it. So we'll try it again. If it's a duplicate, forgive me.

Commit:
Cleared Fatal crash if JTestConfig not found

Added different config file to JDatabaseMySQLTest to allow an independent decision
by the tester on whether these tests should be run. It now requires config_mysql.php
to contain the JMySQLTestConfig class to be present or the tests will be skipped.

Commit:
Skipped Bad Tests

There is a bug in the system strcoll routine either in Darwin or in all BSD
systems that causes locale to be ignored, so arranged for all tests involving
strcoll with a locale on Darwin to be skipped, as their results would only be
coincidentally accurate.

eddieajau commented on tests/suite/joomla/database/database/JDatabaseMySQLTest.php:
```diff
@@ -59,7 +59,10 @@ class JDatabaseMySQLTest extends JoomlaDatabaseTestCase
 	{
-		$config = new JTestConfig;
+		@include_once JPATH_TESTS . '/config.php';
+		if (class_exists('JTestConfig')) {
```
I'm not sure I understand what is gained. If JTestConfig does not exist we are still going to get an error, no?

Paladin replied:
Yes, but....

The severity of the error changes. The original code crashed my testing system completely at this point with a PHP Fatal error if the config.php file wasn't there; no further tests were executed. This code merely errors out of the tests in this file while letting all the subsequent tests in the suite execute.

Long term it's probably not the best solution. I was putting this code in place as a stopgap, to eliminate the crashes so I could see the "big picture" of the testing. It's the same code someone else used in JLoggerDatabaseTest. 

If you want a different fix for it, you might add this else clause at line 66 to the class_exists test (both in this file and for that matter in JLoggerDatabaseTest):

``` php
else
{
    $this->markTestSkipped("These tests require a test config file be present, and I can't find yours.");
}
```

which will still keep the db tests from running but also kill the "Undefined variable" message for these tests. I decided against this idea because the error message, while not blocking further test execution, calls attention to a problem in the setup of the testing system. An even better solution might be to come up with a "standard" default set of config params and use them when the test config isn't present. I'd planned on addressing that question later, after dealing with the unwritten tests; this was simply a means to get the all the tests to run in the meantime.

pasamio replied:
First of all some times you want to test without necessarily having the database tests run. I'd err on marking it as skipped. The other issue is what happens when you shift to Oracle/MSSQL/sqlite3 to do testing? The test should at least check that the driver in the test config is MySQL before continuing.

Paladin replied:
Shows when I last looked at the drivers. Last time I checked there were only mySQL drivers available. I'm not sure how to parse "I'd err on marking it as skipped" -- you mean you'd rather it was an error than a skip? (I agree, BTW.) As for setting up separate suites with and without db tests, that's beyond the scope of what I was doing. (I'm following the principle of test first, pretty later. You want to create those suites, feel free. I see a few thousand tests I'm going to have to write yet before then.)

Commit:
JObservableTest

Added JObservableTester.php
Wrote unit tests for this class, using an extended version of the class to set a protected
attribute for one test and mocking JObserver for the rest. Fixed missing argument in notify
method by sending the observed object in array as 'subject'.

LouisLandry commented on tests/suite/joomla/application/cli/JDaemonTest.php:
```diff
@@ -33,6 +33,9 @@ class JDaemonTest extends PHPUnit_Framework_TestCase
 	 */
 	public function testWriteProcessIdFile()
 	{
+		if (!defined('SIGHUP')) {
```
I believe this should be covered now with a separate pull request that was merged.  If you update the platform and look at line 24 of this file you will find a conditional in the setup method that marks the test skipped if the PCNTL extension isn't loaded.  Can you verify that this problem is fixed with that change and if so we can pass on this addition.  I feel it is a more complete solution given that nothing in JDaemon will work without PCNTL.

Paladin replied:
Feel free to reject this commit, then, if you've put a better solution in the tree since I wrote it.

Commit:
cache file deleted

Commit:
Cleaned up after methods were removed from JDocument and JDocumentJSONTest without
removing associated tests.
